stl/inc/semaphore

```diff
@@ -0,0 +1,304 @@
+_STL_DISABLE_CLANG_WARNINGS
+#pragma push_macro("new")
+#undef new
+
+_STD_BEGIN
+
+template <class _Rep, class _Period>
+_NODISCARD unsigned long long _Semaphore_deadline(const chrono::duration<_Rep, _Period>& _Rel_time) {
+    return __std_atomic_wait_get_deadline(
+        chrono::duration_cast<chrono::duration<unsigned long long, milli>>(_Rel_time).count());
+}
+
+template <class _Clock, class _Duration>
+_NODISCARD unsigned long _Semaphore_remaining_timeout(const chrono::time_point<_Clock, _Duration>& _Abs_time) {
+    auto _Now = _Clock::now();
+    if (_Now >= _Abs_time) {
+        return 0;
+    }
+
+    auto _Rel_time = _Abs_time - _Now;
+    static constexpr chrono::hours _Ten_days{24 * 10};
+    if (_Rel_time >= _Ten_days) {
+        return chrono::duration_cast<chrono::duration<unsigned long, milli>>(_Ten_days).count();
+    } else {
+        return chrono::duration_cast<chrono::duration<unsigned long, milli>>(_Rel_time).count();
+    }
+}
+
+inline constexpr ptrdiff_t _Semaphore_max = (1ULL << (sizeof(ptrdiff_t) * CHAR_BIT - 1)) - 1;
+
+template <ptrdiff_t _Least_max_value = _Semaphore_max>
+class counting_semaphore {
+public:
+    _NODISCARD static constexpr ptrdiff_t(max)() noexcept {
+        return _Least_max_value;
+    }
+
+    constexpr explicit counting_semaphore(const ptrdiff_t _Desired) noexcept /* strengthened */
+        : _Counter(static_cast<ptrdiff_t>(_Desired)) {
+        _STL_VERIFY(_Desired >= 0 && _Desired <= _Least_max_value,
+            "Precondition: desired >= 0, and desired <= max() (N4861 [thread.sema.cnt]/5)");
+    }
+
+    counting_semaphore(const counting_semaphore&) = delete;
+    counting_semaphore& operator=(const counting_semaphore&) = delete;
+
+    void release(const ptrdiff_t _Update = 1) noexcept /* strengthened */ {
+        if (_Update == 0) {
+            return;
+        }
+        _STL_VERIFY(_Update > 0 && _Update <= _Least_max_value,
+            "Precondition: update >= 0, and update <= max() - counter (N4861 [thread.sema.cnt]/8)");
+
+        // We need to notify (wake) at least _Update waiting threads.
+        // Errors towards waking more cannot be always avoided, but they are performance issue.
+        // Errors towards waking less should be avoided, as they are correctness issue.
+
+        // release thread: Increment semaphore counter, then load waiting counter;
+        // acquire thread: Increment waiting counter, then load semaphore counter;
+
+        // memory_order_seq_cst for all four operations guarantees that either release thread loads
+        // the incremented value, or acquire thread loads the incremented value, or both, but not none.
+        // memory_order_seq_cst might be superfluous for some hardware mappings of C++ memory model,
+        // but from the point of C++ memory model itself it is needed, weaker orders don't work.
+
+        const ptrdiff_t _Prev = _Counter.fetch_add(static_cast<ptrdiff_t>(_Update), memory_order_seq_cst);

```

Comment:
I don't think the seq_cst ones should have a `memory_order` supplied since that pessimizes debug mode and doesn't really add clarity.